### PR TITLE
feat(column-entity) Add an alias to entity and column 

### PR DESCRIPTION
**Pull request by evanh**
Add an alias field to an entity, and add an entity field to a column. This is in
preparation for adding JOIN support. Having Column dependent on Entity forced a
code restructure which was huge as everything uses Column. Move all complex
expressions into their own modules.

### CODE REVIEW

File: snuba_sdk/entity.py [hunk 1]
```diff
@@ -15,12 +15,13 @@ class InvalidEntity(Exception):
 @dataclass(frozen=True)
 class Entity(Expression):
     name: str
+    alias: Optional[str] = None
```
Comment (evanh) [inline snuba_sdk/entity.py:18]: This is another functional change

File: snuba_sdk/query_visitors.py [hunk 2]
```diff
@@ -341,10 +339,7 @@ def _combine(self, query: "main.Query", returns: Mapping[str, None]) -> None:
         # - Must have certain conditions (project, timestamp, organization etc.)
         ## SUBQUERIES
         # - outer query must only reference columns from inner query, and reference by alias
-        # - inner query must be valid
         if isinstance(query.match, main.Query):
-            self.visit(query.match)
```
Comment (fpacifici) [inline snuba_sdk/query_visitors.py:346]: Why this change ?
Moreover, shouldn't we validate that, if the query is not a join entity alias must not be used?

Comment (evanh) [reply snuba_sdk/query_visitors.py:346]: This is doing duplicate work. The combine function is called after the `_visit_match` function so the query will already be validated.

Comment (evanh) [reply snuba_sdk/query_visitors.py:346]: The validation comment is valid, I'll check that.

Comment (evanh) [reply snuba_sdk/query_visitors.py:346]: So instead of adding validation I decided to add a flag to the translator that will omit the entity aliases for non-join queries. The reason why I did this was to avoid the end user having to have different entity/column values (mostly entity) when they are building join vs. non-join queries. This way if the end user wants to have a static set of Entities that get reused in different queries, they only need to have one set no matter which query they are building.

File: snuba_sdk/visitors.py [hunk 3]
```diff
@@ -153,8 +151,16 @@ def _visit_debug(self, debug: Debug) -> TVisited:
 
 
 class Translation(ExpressionVisitor[str]):
+    def __init__(self, use_entity_aliases: bool = False):
+        # Eventually JOINs will set this to True, but single entity/sub queries
+        # don't support entity aliases.
+        self.use_entity_aliases = use_entity_aliases
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba_sdk/visitors.py:157]: This is probably better addressed in the Join PR (https://github.com/getsentry/snuba-sdk/pull/35).

> Isn't this going to miss the validation rule that all columns in the joins need to reference a table alias?

No in the Join PR I add explicit validation for that.

> This way we should be able to validate different types of queries cleanly and without letting a simple query get into a state that is undefined for it or the join query to get into an invalid state.

I'm not sure how this would prevent that now. Queries in general (not just Join queries) can already exist in an invalid state, since they can be built up over time. They are not expected to be valid until `query.validate()` is called. I did investigate doing the split you are talking about, but ultimately the difference in validation ended up being just a few extra lines of code.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba_sdk/visitors.py:157]: Sorry I missed the gap between this file and the previous one. I thought this was the query translator while it was the expression one. Yeah my suggestion would not help with the validation rule for joins.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba_sdk/visitors.py:157]: > I did investigate doing the split you are talking about, but ultimately the difference in validation ended up being just a few extra lines of code.

Ok.

**Review by fpacifici — APPROVED** [review]
